### DISCUSSION TITLE
Delete product images

### DISCUSSION THREAD

**Discussion by bacarybruno** [medusajs/medusa#5149] [Feature Requests]
Hello 👋 As suggested by @olivermrbl, I'm opening this feature request regarding product image deletion.

### Current behaviour
Currently, product thumbnail and images are not deleted when the product is removed. The file services `delete` method is not called when the image is removed.

This issue is referenced in:
- https://github.com/medusajs/medusa/issues/1468
- https://github.com/medusajs/medusa/issues/1872
- https://github.com/medusajs/medusa/issues/3774
- https://github.com/medusajs/medusa/issues/4092

### Feature Request
Having the ability to delete product images when a product is removed. This includes:
- Deleting an image when it's removed from the UI
- Deleting all images when the product is removed
- Deleting the thumbnail when it's deleted from the UI
- Deleting the thumbnail when it's replaced in the UI

In this PR https://github.com/medusajs/medusa/pull/5129 I'm calling the fileService to delete the images when the product is updated.
This should be explicitly enabled via a medusa config, so that existing users won't be impacted, and also because product is soft removed so some users might prefer to keep the files.

My first approach was to listen to product events but they don't include old state, so implementing the logic to remove the files was not possible. It would also require to implement the same behaviour in all the files plugins.

What would be the best approach to implement this feature ? 🙏

**Comment by genox** [comment]
I would appreciate the option to select whether attached files should be deleted on deleting a product. Could be a simple checkbox on the delete confirmation dialog.

**Reply by bacarybruno** [reply]
That's a good idea @genox 💡
What could be done for the images that are removed from the UI ? Do you think we should also include a checkbox for that case ?

**Reply by genox** [reply]
I had a look at Postgres tables and realised that products are never really deleted but end up in a "trash can state" that isn't represented in the UI. In this case, deleting the images might have consequences that are not clear yet. Are they supposed to be able to be reinstantiated at some point down the line? In this case, deleting the images might not be what is best.

Maybe it would be better to handle this separately in the "Edit Media" section.

**Reply by genox** [reply]
<img width="810" alt="Screenshot 2023-09-22 at 08 46 24" src="https://github.com/medusajs/medusa/assets/5611407/0e780b53-c4f3-49a9-b116-405f91622404">
<img width="431" alt="Screenshot 2023-09-22 at 08 46 13" src="https://github.com/medusajs/medusa/assets/5611407/35b16816-63a2-435c-81c6-567c5638c2ca">

I think if a user decides to "delete" an image in those sections, the file should be removed from storage. They can't be used anywhere else. No checkbox necessary - there is no way to bring them back and re-attach them to the product or variant anymore, anyways. That alone should prevent file storages to grow with orphaned images substantially.

**Comment by olivermrbl** [comment]
This will affect places where product images are "snapshotted" e.g. on line items.

So for example:
- Add to cart a product with thumbnail
- Product thumbnail updated -> previous thumbnail is deleted from file storage
- Originally added line item will contain nonexisting image

Another example are notifications, such as emails.

I don't find this to be a huge issue, as it would only happen very rarely. Also, in the end, this will not "hurt" the customer, only slightly degrade their experience. 

Something to be wary of.

**Reply by genox** [reply]
That's true. All existing references will be broken.

I see your point tho - for the project I am currently working on, images amount to around 1200 files @350mb total when resized to some sane screen size and jpeg compressed. Even when the entire catalogue is changed once, it still would only be 700mb. In the grand scheme of things ... not much. But in 3-4 years after 6-8 seasonal catalogue updates, backups will contain a lot of data that is technically stale and the % amount of EOL products will always increase.

**Reply by tsvetann** [reply]
I think we can provide a balanced solution for everyone. Emit an event when media is deleted so everyone can choose to ignore or implement a subscriber to remove the media. Everybody Happy! What do you think @olivermrbl ?

**Reply by pevey** [reply]
I think this is still not great UX, because the admin makes it appear as though the image is actually being deleted when it is not.

There are some other enhancements to how images are handled that could be tackled at the same time, and might provide a compromise.

In many other ecommerce systems, there is a media gallery where you upload assets.  You can also delete from here, but there will be a warning if the image is currently attached to any products/categories.  On a product page, you can then select and associate images.  This can be hugely helpful if you want to share certain images across an entire group of products.  You can also "dissociate" or "remove" images, but from here you don't actually delete them.  You can also upload images here, and it will add it the media library automatically.

The images table already has all the data needed to add this to the Admin.

**Reply by genox** [reply]
> I think this is still not great UX, because the admin makes it appear as though the image is actually being deleted when it is not.

Yeah, this is what threw me off and I was about to write it off as a bug.

Why not make permanent media file deletion opt-in?

Since not everybody is using the same frontend specs and has different requirements, it could be just left to the shop owner's preferences wether or not they want to permanently archive images (e.g. leave them as is) or delete them when products are deleted or when they are (seemingly) deleted in the UI.

Speaking for myself, I wouldn't mind showing placeholders for images that can't be retrieved. This would mean the product is most likely EOL anyways. I think not even Zalando keeps EOL product images around forever. At least I think I remember the occasional placeholder when browsing old orders. ;)

One might even consider doing a search/replace over line items and scrub those images when the file is being deleted, to make sure there are no unnecessary references.

**Reply by genox** [reply]
> .. media gallery ..

Some headless CMS do that and while it is helpful in some cases, I think the reality is that media files are mostly only relevant for one content item. More so the fact that after using a system for a while and adding a lot of content, the media library might grow to an ungodly amount of items of which a minimum is actually important for the content being currently edited.

Sanity.io does that for example and it always bugged me that when I upload an image I want to use on content item X, I see 400 images I never want to use in the current context. For the few occasions where I could actually reuse an already uploaded image, it would have made basically no difference to have a duplicate - it would amount to very little overhead and furthermore reduce cross dependencies between content items.

Of course, this is again a question of the use case. I am actually quite happy to handle media files isolated per product. But YMMV, as they say :)

**Reply by bacarybruno** [reply]
Is there an easy way to check all the places where a given image is referenced?
This would help either informing the customer before the deletion, or maybe replacing images with a placeholder as @genox said. Except maybe notifications.

**Reply by tsvetann** [reply]
Good points. I think the bad DX comes from the fact that an admin has to click on a trash icon (which implies physicial deletion of image)

**Comment by pevey** [comment]
> Why not make permanent media file deletion opt-in?

This is probably the sane approach, and could be done via a UX change rather than an overall setting.  The "Delete" button label could be changed to "Remove," and when pressed could present a modal option to delete the file as well, yes or no.
